### PR TITLE
created multithreaded server, updated ClientServerTest

### PR DESCRIPTION
**Pull request by Jovan-Stepanovic**

### CODE REVIEW

File: src/main/java/org/jstepanovic/server/service/Server.java [hunk 1]
```diff
@@ -6,50 +6,64 @@
...
 
 
-public class Server {
+public enum Server {
+    INSTANCE;
 
-    private final int numberOfProcessors;
-
-    private static Server instance;
+    private final int numberOfProcessors = Runtime.getRuntime().availableProcessors();
```
Comment (sigmoidmet) [inline src/main/java/org/jstepanovic/server/service/Server.java:20]: just small comment here: usually you want to use max(availableProcessors - 1, 1) for long-running servers to leave some space for garbage collector and other system threads

File: src/main/java/org/jstepanovic/server/service/Server.java [hunk 2]
```diff
@@ -6,50 +6,64 @@
...
+    private final ExecutorService executorService = Executors.newFixedThreadPool(numberOfProcessors);
 
     private final Repository repository = new Repository();
 
-    private Server() {
-        numberOfProcessors = Runtime.getRuntime().availableProcessors();
-    }
+    private boolean isServerRunning;
```
Comment (sigmoidmet) [inline src/main/java/org/jstepanovic/server/service/Server.java]: If you will try to read this variable in different threads you may see some awkward results. You may read about happens-before
Caution, this one is pretty long and mind-blowing:)
So maybe you want to find something else on this topic
[Java Memory Model Pragmatics](https://shipilev.net/blog/2014/jmm-pragmatics/)

File: src/main/java/org/jstepanovic/server/service/Server.java [hunk 3]
```diff
@@ -6,50 +6,64 @@
...
+            executorService.submit(new Session(serverSocket.accept(), repository));
 
-            executorService.shutdown();
+            while (true) {
+                try {
+                    Socket clientSocket = serverSocket.accept();
+                    System.out.println("started communication with the new client");
+                    executorService.submit(new Session(clientSocket, repository));
+                } catch (SocketException e) {
+                    isServerRunning = false;
```
Comment (sigmoidmet) [inline src/main/java/org/jstepanovic/server/service/Server.java]: I'd suggest to change this variable only in stop method and call it directly if server needs to be shut down due to some error for example.

File: src/main/java/org/jstepanovic/server/service/Server.java [hunk 4]
```diff
@@ -6,50 +6,64 @@
...
+                }
+            }
 
-        } catch (IOException | ExecutionException | InterruptedException e) {
+        } catch (IOException e) {
             e.printStackTrace();
         }
     }
+
+    public synchronized void stop() {
```
Comment (sigmoidmet) [inline src/main/java/org/jstepanovic/server/service/Server.java:54]: You may call this method several times, it's good to have some check here if you need an actual stop or it was done before.

File: src/main/java/org/jstepanovic/server/service/Server.java (hunk 1 above)
Comment (Jovan-Stepanovic) [reply src/main/java/org/jstepanovic/server/service/Server.java:20]: Done

File: src/main/java/org/jstepanovic/server/service/Server.java (hunk 2 above)
Comment (Jovan-Stepanovic) [reply src/main/java/org/jstepanovic/server/service/Server.java]: using AtomicBoolean

File: src/main/java/org/jstepanovic/server/service/Server.java (hunk 3 above)
Comment (Jovan-Stepanovic) [reply src/main/java/org/jstepanovic/server/service/Server.java]: refactored

File: src/main/java/org/jstepanovic/server/service/Server.java (hunk 4 above)
Comment (Jovan-Stepanovic) [reply src/main/java/org/jstepanovic/server/service/Server.java:54]: refactored

**Review by sigmoidmet — APPROVED** [review]
LGTM

**Pull request merged**